pkg/server/commands/listusers/list_users_rpc.go

```diff
@@ -658,20 +665,15 @@ func (l *listUsersQuery) expandExclusion(
 	wildcardKey := tuple.TypedPublicWildcard(req.GetUserFilters()[0].GetType())
 
 	_, baseWildcardExists := baseFoundUsersMap[wildcardKey]
-	subtractWildcardUser, subtractWildcardExists := subtractFoundUsersMap[wildcardKey]
-
-	if baseWildcardExists && (!subtractWildcardExists || subtractWildcardUser.relationshipStatus == NoRelationship) {
-		trySendResult(ctx, foundUser{
-			user: tuple.StringToUserProto(wildcardKey),
-		}, foundUsersChan)
-	}
+	_, subtractWildcardExists := subtractFoundUsersMap[wildcardKey]
 
 	for userKey, fu := range baseFoundUsersMap {
 		subtractedUser, userIsSubtracted := subtractFoundUsersMap[userKey]
 		_, wildcardSubtracted := subtractFoundUsersMap[wildcardKey]
 
-		if baseWildcardExists {
-			if !userIsSubtracted && !wildcardSubtracted {
+		switch {
+		case baseWildcardExists:
+			if !wildcardSubtracted {
```

Comment:
This conditional behavior is not quite correct and has deviated from the prior commit. Previously we were only producing the wildcard subject if `baseWildcardExists && !wildcardSubtracted`. See https://github.com/openfga/openfga/blob/b872f91c05f8c130e369656e4bddf8dc3ac47914/pkg/server/commands/listusers/list_users_rpc.go#L665

Consider the following scenario:

```
type user

type document
  relations
    define restricted: [user, user:*]
    define viewer: [user, user:*] but not restricted
```
with tuples:
* document:1#viewer@user:*
* document:1#viewer@user:jon
* document:1#restricted@user:jon

When ranging over the `baseFoundUsersMap` we'll find `user:jon` and yield it here, because there is not a subtracted wildcard. The only reason a test related to this scenario is passing against this commit is because of other code in the callstack above this function, which isn't obvious and is actually misleading when reading the code. See https://github.com/openfga/openfga/blob/5f20ffb21dfee1e6aef05b6360dd4e8208bf3425/pkg/server/commands/listusers/list_users_rpc.go#L251-L253

But this logic itself is not quite correct here. We're producing on the `foundUsersChan` superfluously only to be filtered out above upstream since it is not applicable (the subject `user:jon` is not just excluded but is explicitly subtracted). Ideally this rewrite handler should only produce accurate "candidates" with respect to the found users channel.

Reply:
Are you ok with proving your above theory with a unit test? Then we can address appropriately in code.

Reply:
https://github.com/openfga/openfga/pull/1604/files#r1606950464